README.md
request
=======
[![Build Status](https://travis-ci.org/mozillazg/request.svg?branch=master)](https://travis-ci.org/mozillazg/request)
[![Coverage Status](https://coveralls.io/repos/mozillazg/request/badge.png?branch=master)](https://coveralls.io/r/mozillazg/request?branch=master)
[![Go Report Card](https://goreportcard.com/badge/github.com/mozillazg/request)](https://goreportcard.com/report/github.com/mozillazg/request)
[![GoDoc](https://godoc.org/github.com/mozillazg/request?status.svg)](https://godoc.org/github.com/mozillazg/request)

A developer-friendly HTTP request library for Gopher. Inspired by [Python-Requests](https://github.com/kennethreitz/requests).


Installation
------------

```
go get -u github.com/mozillazg/request
```


Documentation
--------------

API documentation can be found here:
https://godoc.org/github.com/mozillazg/request


Usage
-------

```go
import (
	"github.com/mozillazg/request"
)
```

**GET**:

```go
c := new(http.Client)
req := request.NewRequest(c)
resp, err := req.Get("http://httpbin.org/get")
j, err := resp.Json()
defer resp.Body.Close()  // Don't forget close the response body
```

**POST**:

```go
req = request.NewRequest(c)
req.Data = map[string]string{
	"key": "value",
	"a":   "123",
}
resp, err := req.Post("http://httpbin.org/post")
```

**Cookies**:

```go
req = request.NewRequest(c)
req.Cookies = map[string]string{
	"key": "value",
	"a":   "123",
}
resp, err := req.Get("http://httpbin.org/cookies")
```

**Headers**:

```go
req = request.NewRequest(c)
req.Headers = map[string]string{
	"Accept-Encoding": "gzip,deflate,sdch",
	"Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
}
resp, err := req.Get("http://httpbin.org/get")
```

**Files**:

```go
req = request.NewRequest(c)
f, err := os.Open("test.txt")
req.Files = []request.FileField{
	request.FileField{"file", "test.txt", f},
}
resp, err := req.Post("http://httpbin.org/post")
```

**Json**:

```go
req = request.NewRequest(c)
req.Json = map[string]string{
	"a": "A",
	"b": "B",
}
resp, err := req.Post("http://httpbin.org/post")
req.Json = []int{1, 2, 3}
resp, err = req.Post("http://httpbin.org/post")
```

**Proxy**:
```go
req = request.NewRequest(c)
req.Proxy = "http://127.0.0.1:8080"
// req.Proxy = "https://127.0.0.1:8080"
// req.Proxy = "socks5://127.0.0.1:57341"
resp, err := req.Get("http://httpbin.org/get")
```
or https://github.com/mozillazg/request/tree/develop/_example/proxy

**HTTP Basic Authentication**:
```go
req = request.NewRequest(c)
req.BasicAuth = request.BasicAuth{"user", "passwd"}
resp, err := req.Get("http://httpbin.org/basic-auth/user/passwd")
```


License
---------

Under the MIT License.


CHANGELOG.md
# Changelog


## 0.8.0 (2016-mm-dd)

* Add `Request.Hooks` for supports before and after request hooks (via #26)


## 0.7.0 (2016-11-07)

* Add auto decompress deflated body (via [#23][#23]. thanks [@DavidCai1993][@DavidCai1993])


## 0.6.0 (2016-09-04)

* Remove vendor/, use glide instead

### API Changes

* Add `var DefaultClient = new(http.Client)`
* Support `NewArgs(nil)`
* Support `NewRequest(nil)`
* Support `GET(url, nil), POST(url, nil), ...`


## 0.5.1 (2016-06-05)

* Add vendor/


## 0.5.0 (2015-11-15)

### API Changes

* Add `func (req *Request) PostForm(url interface{}, data interface{}) (resp *Response, err error)`
* Add `func (req *Request) Reset()`


## 0.4.0 (2015-10-31)

### API Changes

* `type Request struct` add `Body` field
* Add new variables:
  * `DefaultHeaders`
  * `DefaultContentType`
  * `DefaultJsonType`
  * `DefaultRedirectLimit`
  * `DefaultUserAgent`

## 0.3.1 (2015-09-24)

* Updated docs
* Improve code
* No API change, No bugfix

## 0.3.0 (2015-02-11)

### API Changes

* Add `type Request struct`, recommend use it instead of `Args`
* Add `func NewRequest(c *http.Client) *Request`
* Add `func (req *Request) Get(url interface{}) (resp *Response, err error)`
* Add `func (req *Request) Head(url interface{}) (resp *Response, err error)`
* Add `func (req *Request) Post(url interface{}) (resp *Response, err error)`
* Add `func (req *Request) Put(url interface{}) (resp *Response, err error)`
* Add `func (req *Request) Patch(url interface{}) (resp *Response, err error)`
* Add `func (req *Request) Delete(url interface{}) (resp *Response, err error)`
* Add `func (req *Request) Options(url interface{}) (resp *Response, err error)`


## 0.2.0 (2015-01-15)

### API Changes

* Add `Args.Proxy` for support HTTP/HTTPS/SOCKS5 proxy
* Add `Args.BasicAuth` for Support HTTP Basic Authentication
* Add `func (resp *Response) URL() (*url.URL, error)`
* Add `var DefaultRedirectLimit = 10`

### Bugfixes

* Fix "http.Client don't use original Header when it do redirect" [#6](https://github.com/mozillazg/request/issues/6)


## 0.1.0 (2015-01-08)

* Initial Release

[#23]: https://github.com/mozillazg/request/pull/23
[@DavidCai1993]: https://github.com/DavidCai1993
[#26]: https://github.com/mozillazg/request/pull/26


doc.go
// Go HTTP Requests for Humans™.
//
// GET Request:
//
// 	c := &http.Client{}
// 	req := request.NewRequest(c)
// 	resp, err := req.Get("http://httpbin.org/get")
// 	defer resp.Body.Close()  // **Don't forget close the response body**
// 	j, err := resp.Json()
//
// POST Request:
//
// 	req = request.NewRequest(c)
//	req.Data = map[string]string{
//		"key": "value",
//		"a":   "123",
//	}
//	resp, err := req.Post("http://httpbin.org/post")
//
// Custom Cookies:
//
// 	req = request.NewRequest(c)
//	req.Cookies = map[string]string{
//		"key": "value",
//		"a":   "123",
//	}
//	resp, err := req.Get("http://httpbin.org/cookies")
//
//
// Custom Headers:
//
// 	req = request.NewRequest(c)
//	req.Headers = map[string]string{
//		"Accept-Encoding": "gzip,deflate,sdch",
//		"Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
//	}
//	resp, err := req.Get("http://httpbin.org/get")
//
// Upload Files:
//
// 	req = request.NewRequest(c)
//	f, err := os.Open("test.txt")
//	req.Files = []request.FileField{
//		request.FileField{"file", "test.txt", f},
//	}
//	resp, err := req.Post("http://httpbin.org/post")
//
// Json Body:
//
// 	req = request.NewRequest(c)
//	req.Json = map[string]string{
//		"a": "A",
//		"b": "B",
//	}
//	resp, err := req.Post("http://httpbin.org/post")
//	req.Json = []int{1, 2, 3}
//	resp, err = req.Post("http://httpbin.org/post")
//
// others body:
//
// 	req = request.NewRequest(c)
//	// not set Content-Type
//	req.Body = strings.NewReader("<xml><a>abc</a></xml")
//	resp, err := req.Post("http://httpbin.org/post")
//
//	// form
// 	req = request.NewRequest(c)
//	req.Body = strings.NewReader("a=1&b=2")
//	req.Headers = map[string]string{
//		"Content-Type": request.DefaultContentType,
//	}
//	resp, err = req.Post("http://httpbin.org/post")
//
// Proxy:
//
// 	req = request.NewRequest(c)
//	req.Proxy = "http://127.0.0.1:8080"
//	// req.Proxy = "https://127.0.0.1:8080"
//	// req.Proxy = "socks5://127.0.0.1:57341"
//	resp, err := req.Get("http://httpbin.org/get")
//
// HTTP Basic Authentication:
//
// 	req = request.NewRequest(c)
//	req.BasicAuth = request.BasicAuth{"user", "passwd"}
//	resp, err := req.Get("http://httpbin.org/basic-auth/user/passwd")
//
// Need more control?
//
// You can setup req.Client(you know, it's an &http.Client),
// for example: set timeout
//
//	timeout := time.Duration(1 * time.Second)
//	req.Client.Timeout = timeout
//	req.Get("http://httpbin.org/get")
package request
